lib/Dialect/Arc/Transforms/LowerToLLVM.cpp

```diff
@@ -0,0 +1,471 @@
+                  ConversionPatternRewriter &rewriter) const override {
+    rewriter.replaceOpWithNewOp<func::ReturnOp>(op, adaptor.getOperands());
+    return success();
+  }
+};
+
+} // namespace
+
+static bool isArcType(Type type) {
+  return type.isa<StorageType>() || type.isa<MemoryType>() ||
+         type.isa<StateType>();
+}
+
+static bool hasArcType(TypeRange types) {
+  return llvm::any_of(types, isArcType);
+}
+
+static bool hasArcType(ValueRange values) {
+  return hasArcType(values.getTypes());
+}
+
+template <typename Op>
+static void addGenericLegality(ConversionTarget &target) {
+  target.addDynamicallyLegalOp<Op>([](Op op) {
+    return !hasArcType(op->getOperands()) && !hasArcType(op->getResults());
+  });
+}
+
+static void populateLegality(ConversionTarget &target) {
+  target.addLegalDialect<mlir::BuiltinDialect>();
+  target.addLegalDialect<hw::HWDialect>();
+  target.addLegalDialect<comb::CombDialect>();
+  target.addLegalDialect<func::FuncDialect>();
+  target.addLegalDialect<scf::SCFDialect>();
+  target.addLegalDialect<LLVM::LLVMDialect>();
+
+  target.addIllegalOp<arc::DefineOp>();
+  target.addIllegalOp<arc::OutputOp>();
+  target.addIllegalOp<arc::StateOp>();
+  target.addIllegalOp<arc::ClockTreeOp>();
+  target.addIllegalOp<arc::PassThroughOp>();
+
+  target.addDynamicallyLegalOp<func::FuncOp>([](func::FuncOp op) {
+    auto argsConverted = llvm::none_of(op.getBlocks(), [](auto &block) {
+      return hasArcType(block.getArguments());
+    });
+    auto resultsConverted = !hasArcType(op.getResultTypes());
+    return argsConverted && resultsConverted;
+  });
+  addGenericLegality<func::ReturnOp>(target);
+}
+
+static void populateTypeConversion(TypeConverter &typeConverter) {
+  typeConverter.addConversion([&](StorageType type) {
+    return LLVM::LLVMPointerType::get(IntegerType::get(type.getContext(), 8));
+  });
+  typeConverter.addConversion([&](MemoryType type) {
+    return LLVM::LLVMPointerType::get(
+        IntegerType::get(type.getContext(), type.getStride() * 8));
+  });
+  typeConverter.addConversion([&](StateType type) {
+    return LLVM::LLVMPointerType::get(
+        typeConverter.convertType(type.getType()));
+  });
+  typeConverter.addConversion([](hw::ArrayType type) { return type; });
+  typeConverter.addConversion([](mlir::IntegerType type) { return type; });
+}
+
+static void populateOpConversion(RewritePatternSet &patterns,
+                                 TypeConverter &typeConverter) {
+  auto *context = patterns.getContext();
+  // clang-format off
+  patterns.add<
+    DefineOpLowering,
+    OutputOpLowering,
+    StateOpLowering,
+    ClockTreeOpLowering,
+    PassThroughOpLowering,
+    AllocStorageOpLowering,
+    AllocStateOpLowering,
+    StateReadOpLowering,
+    StateWriteOpLowering,
+    AllocMemoryOpLowering,
+    MemoryReadOpLowering,
+    MemoryWriteOpLowering,
+    ClockGateOpLowering,
+    StorageGetOpLowering,
+    ReturnOpLowering
+  >(typeConverter, context);
```

Comment:
Super-nit: either group them by category with comments inbetween and enable clang-format, or properly sort alphabetically.

Reply:
:innocent: good idea